File: longest subString without repeating characters/hashmap.js

JavaScript

function longestSubStringWithNoRepatingCharacters(s) {
    // Initialize pointers, maximum length, and a map to store character indices
    let left = 0, right = 0, maxLen = 0, longestSubString = '';
    const sLen = s.length;
    const charMap = new Map();

    // Iterate through the string
    while (right < sLen) {
        // If the current character is not in the map or its previous occurrence is before the current left pointer
        if (!charMap.has(s[right]) || charMap.get(s[right]) < left) {
            // Update the character's index in the map
            charMap.set(s[right], right);
            // Move the right pointer to the next character
            right++;
            // Update the maximum length if necessary
            maxLen = Math.max(maxLen, right - left);
            // Update the longest substring if necessary
            longestSubString = s.substring(left, right);
        } else {
            // If the current character is a duplicate, move the left pointer to the position after the previous occurrence
            left = charMap.get(s[right]) + 1;
            // Update the character's index in the map
            charMap.set(s[right], right);
        }
    }

    return [longestSubString, maxLen];
}


const str = "abbccabcd"
const [ longestSubString, maxLen] = longestSubStringWithNoRepatingCharacters(str)
console.log(`String lenght:\t ${maxLen}`);
console.log(`subString:\t ${longestSubString}`);

File: longest subString without repeating characters/Python/hashmap.py
def longest_substring_with_no_repeating_characters(s):
    # Initialize pointers, maximum length, and a dictionary to store character indices
    left = 0  # Left pointer for the sliding window
    max_len = 0  # To store the maximum length of the substring
    longest_substring = ''  # To store the actual longest substring
    char_map = {}  # Dictionary to store the latest index of each character

    # Iterate through the string using the right pointer
    for right, current_char in enumerate(s):
        # If the character is in the map and its index is on or after the left pointer
        if current_char in char_map and char_map[current_char] >= left:
            # Move the left pointer to the position after the previous occurrence
            left = char_map[current_char] + 1

        # Update the character's index in the map
        char_map[current_char] = right

        # Update the maximum length and longest substring if the current window is larger
        if right - left + 1 > max_len:
            max_len = right - left + 1
            longest_substring = s[left:right + 1]

    # Return the longest substring and its length
    return longest_substring, max_len

# Test the function with the given string
str_input = "abbccabcd"
longest_substring, max_len = longest_substring_with_no_repeating_characters(str_input)
print(f"String length:\t {max_len}")
print(f"Substring:\t {longest_substring}")